services/distribution/src/test/java/app/coronawarn/server/services/distribution/assembly/diagnosiskeys/structure/directory/DiagnosisKeysDateDirectoryTest.java

```diff
@@ -0,0 +1,186 @@
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+import org.junit.rules.TemporaryFolder;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.boot.test.context.ConfigFileApplicationContextInitializer;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit.jupiter.SpringExtension;
+
+@EnableConfigurationProperties(value = DistributionServiceConfig.class)
+@ExtendWith(SpringExtension.class)
+@ContextConfiguration(classes = {CryptoProvider.class, DistributionServiceConfig.class},
+    initializers = ConfigFileApplicationContextInitializer.class)
+class DiagnosisKeysDateDirectoryTest {
+
+  @Rule
+  private final TemporaryFolder outputFolder = new TemporaryFolder();
+
+  @Autowired
+  CryptoProvider cryptoProvider;
+
+  @Autowired
+  DistributionServiceConfig distributionServiceConfig;
+
+  private File outputFile;
+
+  private static Stream<Arguments> createDiagnosisKeysAndExpectedFiles() {
+    return Stream.of(
+        //Creates correct structure for multiple dates
+        new TestTuple(IntStream.range(0, 3)
+            .mapToObj(currentDate -> IntStream.range(0, 5)
+                .mapToObj(currentHour ->
+                    buildDiagnosisKeys(6, LocalDateTime.of(1970, 1, 3 + currentDate, 0, 0).plusHours(currentHour), 5))
+                .flatMap(List::stream)
+                .collect(Collectors.toList()))
+            .flatMap(List::stream)
+            .collect(Collectors.toList()),
+            Helpers.getExpectedDateAndHourFiles(Map.of(
+                "1970-01-03", List.of("0", "1", "2", "3", "4"),
+                "1970-01-04", List.of("0", "1", "2", "3", "4"),
+                "1970-01-05", List.of("0", "1", "2", "3", "4")), "1970-01-06"),
+            LocalDateTime.of(1970, 1, 6, 0, 0)
+        ),
+        //Does not include the current date
+        new TestTuple(IntStream.range(0, 3)
+            .mapToObj(currentDate -> IntStream.range(0, 5)
+                .mapToObj(currentHour ->
+                    buildDiagnosisKeys(6, LocalDateTime.of(1970, 1, 3 + currentDate, 0, 0).plusHours(currentHour), 5))
+                .flatMap(List::stream)
+                .collect(Collectors.toList()))
+            .flatMap(List::stream)
+            .collect(Collectors.toList()),
+            Helpers.getExpectedDateAndHourFiles(Map.of(
+                "1970-01-03", List.of("0", "1", "2", "3", "4"),
+                "1970-01-04", List.of("0", "1", "2", "3", "4"),
+                "1970-01-05", List.of("0", "1", "2", "3", "4")), "1970-01-05"),
+            LocalDateTime.of(1970, 1, 5, 12, 0)
+        ),
+        //Does not include empty dates
+        new TestTuple(IntStream.range(0, 3)
```

Comment:
I find this procedural generation of input parameters rather obscure. Maybe it being a 12-lines chained statement also contributes to that.